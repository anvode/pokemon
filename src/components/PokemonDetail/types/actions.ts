import { Pokemon } from './PokemonDetail';

export const POKEMON_DETAIL_LIST = 'POKEMON_DETAIL_LIST';
export const POKEMON_DETAIL_LOADING = 'POKEMON_DETAIL_LOADING';
export const POKEMON_DETAIL_ERROR = 'POKEMON_DETAIL_ERROR';

export interface SetPokemonDetailAction {
    type: typeof POKEMON_DETAIL_LIST;
    payload: {
        pokemonDetail: Pokemon;
        id: number;
    }
}

export interface SetPokemonDetailErrorAction {
    type: typeof POKEMON_DETAIL_ERROR;
    payload: {
        error: Error | null;
    }
}

export interface SetPokemonDetailLoadingAction {
    type: typeof POKEMON_DETAIL_LOADING;
    payload: {
        loading: boolean;
    }
}

export type PokemonDetailActionTypes = SetPokemonDetailErrorAction | SetPokemonDetailLoadingAction | SetPokemonDetailAction;
