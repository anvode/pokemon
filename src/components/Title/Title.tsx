
import * as React from 'react';
import { useSelector } from 'react-redux';

import { RootState } from '../../redux/root-reducer';
            
export interface TitleProps {}
const Title: React.FC<TitleProps> = () => {
    const { title } = useSelector((state: RootState) => state.pokemon);
        
    return <>
        <h1 data-testid="h1" className="app__title bg-light p-3 mb-3">{title}</h1>

    </>;
};
            
export default Title;