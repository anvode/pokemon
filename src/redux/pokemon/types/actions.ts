import { PokemonItem } from './Pokemon';

export const POKEMON_FETCH = 'POKEMON_FETCH';
export const POKEMON_FETCH_LOADING = 'POKEMON_FETCH_LOADING';
export const POKEMON_FETCH_ERROR = 'POKEMON_FETCH_ERROR';

export interface SetPokemonAction {
    type: typeof POKEMON_FETCH;
    payload: {
        pokemonList: PokemonItem[], 
        next: string | null, 
        previous: string | null, 
        offset: number, 
        limit: number,
        count: number
    };
}

export interface SetPokemonErrorAction {
    type: typeof POKEMON_FETCH_ERROR;
    payload: {
        error: boolean;
    }
}

export interface SetPokemonLoadingAction {
    type: typeof POKEMON_FETCH_LOADING;
    payload: {
        loading: boolean;
    }
}

export type PokemonActionTypes = SetPokemonAction | SetPokemonErrorAction | SetPokemonLoadingAction;
