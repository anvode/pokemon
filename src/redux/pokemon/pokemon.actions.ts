import { PokemonItem } from './types/Pokemon';
import { POKEMON_FETCH, POKEMON_FETCH_LOADING, POKEMON_FETCH_ERROR, PokemonActionTypes } from './types/actions';

export const setPokemonAction = (pokemonList: PokemonItem[], next: string | null, previous: string | null, offset: number, limit: number, count: number): PokemonActionTypes => {
    return {
        type: POKEMON_FETCH,
        payload: {
            pokemonList,
            next,
            previous,
            offset,
            limit, 
            count
        }
    };
};

export const setPokemonFetchErrorAction = (error: boolean): PokemonActionTypes => {
    return {
        type: POKEMON_FETCH_ERROR,
        payload: {
            error: error
        }
    };
};

export const setPokemonFetchLoadingAction = (loading: boolean): PokemonActionTypes => {
    return {
        type: POKEMON_FETCH_LOADING,
        payload: {
            loading: loading 
        }

    };
};

