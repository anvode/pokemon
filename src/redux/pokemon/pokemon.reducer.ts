import { PokemonState } from './types/Pokemon';
import { POKEMON_FETCH, POKEMON_FETCH_LOADING, POKEMON_FETCH_ERROR, PokemonActionTypes } from './types/actions';

export const initialState: PokemonState = {
    pokemonFetchLoading: false,
    pokemonFetchError: false,
    pokemonList: [],
    next: null,
    previous: null,
    offset: 0,
    limit: 50,
    count: 0
};

export const pokemonReducer = (state: PokemonState = initialState, action: PokemonActionTypes): PokemonState => {

    switch (action.type) {

        case POKEMON_FETCH: {
            return {
                ...state,
                pokemonList: action.payload.pokemonList,
                next: action.payload.next,
                previous: action.payload.previous,
                offset: action.payload.offset,
                limit: action.payload.limit,
                count: action.payload.count
            };
        }

        case POKEMON_FETCH_LOADING: {
            return {
                ...state,
                pokemonFetchLoading: action.payload.loading
            };
        }

        case POKEMON_FETCH_ERROR: {
            return {
                ...state,
                pokemonFetchError: action.payload.error
            };
        }

        default:
            return state;
    }
};
