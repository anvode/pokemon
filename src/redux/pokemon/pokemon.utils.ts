import { Dispatch } from 'redux';
import { setPokemonAction, setPokemonFetchErrorAction, setPokemonFetchLoadingAction } from './pokemon.actions';

export const fetchPokemon = (offset: number, limit: number) => {

    return async (dispatch: Dispatch): Promise<void> => {
        dispatch(setPokemonFetchLoadingAction(true));

        try {
            const response = await fetch(`${process.env.REACT_APP_POKEMON_API_URL}?offset=${offset}&limit=${limit}`);
            const {results: pokemonList, next, previous, count} = await response.json();
            dispatch(setPokemonAction(pokemonList, next, previous, offset, limit, count));
            dispatch(setPokemonFetchLoadingAction(false));
    
        } catch (err) {
            dispatch(setPokemonFetchLoadingAction(false));
            dispatch(setPokemonFetchErrorAction(err.message));
        }
    };
};

