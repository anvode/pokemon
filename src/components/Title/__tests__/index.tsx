import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider, useDispatch } from 'react-redux';
import Title from '../Title';
import { setTitleAction } from '../../../redux/pokemon/pokemon.actions';

import Store from '../../../redux/store';

describe('App', () => {
    afterEach(cleanup);
    
    it('renders', () => {
        const { asFragment } = render(<Title />, {
            wrapper: function child({ children }) {return <Provider store={Store}><BrowserRouter>{children}</BrowserRouter></Provider>;}
        });
        expect(asFragment()).toMatchSnapshot();
    });

});