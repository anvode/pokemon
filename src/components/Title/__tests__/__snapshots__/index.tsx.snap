// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<DocumentFragment>
  <h1
    class="app__title bg-light p-3 mb-3"
    data-testid="h1"
  />
</DocumentFragment>
`;
